core/src/main/java/org/testng/reporters/FailedReporter.java

```diff
@@ -193,13 +194,27 @@ private void createXmlTest(
       List<XmlInclude> methodNames = Lists.newArrayList(methodList.size());
       int ind = 0;
       for (ITestNGMethod m : methodList) {
-
         XmlInclude methodName =
             new XmlInclude(m.getMethodName(), m.getFailedInvocationNumbers(), ind++);
         methodName.setParameters(findMethodLocalParameters(srcXmlTest, m));
+        methodName.setConstructorOrMethod(m.getConstructorOrMethod());
         methodNames.add(methodName);
       }
-      xmlClass.setIncludedMethods(methodNames);
+
+      List<XmlInclude> groupedByMethodNames = new ArrayList<>();
```

Comment:
@PavelSakharchuk - We can refactor this changeset into something like below (Which would prevent us from injecting a `ConstructorOrMethod` object into `XmlInclude`. Ideally speaking the `XmlInclude` should not be even aware of `ConstructorOrMethod` object.


Here's a refactored implementation for computing groupedByMethodNames (This would also mean that we can get rid of the earlier `for..loop` that attempts to first create the `methodNames` list.


```java
      Map<ConstructorOrMethod, List<ITestNGMethod>> group = methodList.stream()
          .collect(Collectors.groupingBy(ITestNGMethod::getConstructorOrMethod));

      List<XmlInclude> groupedByMethodNames = group.values()
          .stream()
          .map(each -> asXmlIncludes(each, srcXmlTest))
          .map(each -> {
            if (each.size() > 1) {
              XmlInclude tmpMethodName = each.get(0);
              each.stream()
                  .map(XmlInclude::getInvocationNumbers)
                  .reduce((a1, a2) -> {
                    Set<Integer> set = new HashSet<>(a1);
                    set.addAll(a2);
                    return new ArrayList<>(set);
                  }).ifPresent(tmpMethodName::addInvocationNumbers);
              return Collections.singletonList(tmpMethodName);
            }
            return each;
          })
          .flatMap(Collection::stream)
          .collect(Collectors.toList());
```

and the implementation of `asXmlIncludes()` would look like below:

```java
  private static List<XmlInclude> asXmlIncludes(List<ITestNGMethod> methods, XmlTest srcXmlTest) {
    AtomicInteger i = new AtomicInteger(0);
    return methods.stream()
        .map(m -> {
          XmlInclude methodName =
              new XmlInclude(m.getMethodName(), m.getFailedInvocationNumbers(),
                  i.getAndIncrement());
          methodName.setParameters(findMethodLocalParameters(srcXmlTest, m));
          return methodName;
        })
        .collect(Collectors.toList());
  }
```

Reply:
Thanks for your feedback. done